presto-main/src/main/java/com/facebook/presto/operator/scalar/KDistinct.java

```diff
@@ -0,0 +1,114 @@
+import it.unimi.dsi.fastutil.longs.LongArraySet;
+
+import java.lang.invoke.MethodHandle;
+import java.util.Optional;
+
+import static com.facebook.presto.common.type.BigintType.BIGINT;
+import static com.facebook.presto.common.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.metadata.BuiltInTypeAndFunctionNamespaceManager.DEFAULT_NAMESPACE;
+import static com.facebook.presto.operator.scalar.ScalarFunctionImplementationChoice.ArgumentProperty.functionTypeArgumentProperty;
+import static com.facebook.presto.operator.scalar.ScalarFunctionImplementationChoice.ArgumentProperty.valueTypeArgumentProperty;
+import static com.facebook.presto.operator.scalar.ScalarFunctionImplementationChoice.NullConvention.RETURN_NULL_ON_NULL;
+import static com.facebook.presto.util.Reflection.methodHandle;
+
+public final class KDistinct
+        extends SqlScalarFunction
+{
+    public static final KDistinct K_DISTINCT = new KDistinct();
+    private static final MethodHandle STATE_FACTORY = methodHandle(KDistinct.class, "createState");
+
+    private KDistinct()
+    {
+        super(new Signature(
+                QualifiedObjectName.valueOf(DEFAULT_NAMESPACE, "topk_distinct"),
+                FunctionKind.SCALAR,
+                ImmutableList.of(),
+                ImmutableList.of(),
+                BOOLEAN.getTypeSignature(),
+                ImmutableList.of(BIGINT.getTypeSignature(), BIGINT.getTypeSignature()),
+                false));
+    }
+
+    @Override
+    public SqlFunctionVisibility getVisibility()
+    {
+        return SqlFunctionVisibility.PUBLIC;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return false;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "add a hash (long) if it is not already seen and k elements are not seen";
+    }
+
+    @Override
+    public BuiltInScalarFunctionImplementation specialize(BoundVariables boundVariables, int arity, FunctionAndTypeManager functionAndTypeManager)
+    {
+        return new BuiltInScalarFunctionImplementation(
+                false,
+                ImmutableList.of(
+                        valueTypeArgumentProperty(RETURN_NULL_ON_NULL),
+                        functionTypeArgumentProperty(BinaryFunctionInterface.class)),
+                methodHandle(KDistinct.class, "addKey", Object.class, long.class, long.class),
+                Optional.of(methodHandle(KDistinct.class, "createState")));
+    }
+
+    @UsedByGeneratedCode
+    public static boolean addKey(Object state, long hash, long size)
+    {
+        State currentState = (State) state;
+        if (currentState.longArraySet != null) {
+            if (currentState.longArraySet.size() == size) {
+                return false;
+            }
+        }
+        else {
+            currentState.longArraySet = new LongArraySet(new long[(int) size]);
+        }
+
+        // new element added so remove the largest one
```

Comment:
I'm not sure I understand how this code is removing "the largest one". I think this code either adds "hash" (if not exists already) or not. It returns true if added and false if the value already existed.

Reply:
Sorry bad comment. Removed it.